### ISSUE TITLE
Adding BS3 (Twitter Bootstrap) support 

### ISSUE THREAD

**Issue by kematzy** [jeremyevans/forme#12]
Hi Jeremy,

Forme looks great, but I think it lacks BS3 (bootstrap) formatting support.

I'm wiling to help with this, but I'm unsure (confused) about the best way to add this support.

Any pointers would be greatly appreciated.

**Comment by jeremyevans** [comment]
If you wanted to add this, it would probably be best to add transformers that handle adding the extra markup bootstrap likes.  Then you group those transformers into a configuration, and then users can select that configuration to get that group of transformers for a form, or even set the bootstrap configuration as the default configuration.

As I use bootstrap on most of my sites, I'd certainly be in favor of adding this.  I just don't want to change the current default behavior.

**Comment by kematzy** [comment]
Hi Jeremy, 

Thanks for your reply.  

I have begun work on this here [https://github.com/kematzy/forme/tree/bs3-support]. It's very rough work-in-progress, but some parts works well.

I would really appreciate if you could find the time to review the code and correct me if I'm going down the wrong path, as I don't fully understand your code.

The spec's are a bit rough, as I don't have any more time today, but I'm sure you get the gist of what I'm trying to do. Please tell me if I'm doing something silly there, as I'm more used to RSpec.

Thanks a lot for your time and attention. Very much appreciated.

**Comment by jeremyevans** [comment]
I don't have time to do a thorough review at present (I'm at an OpenBSD hackathon), but you are definitely on the right path.  The only thing i noticed is you probably want to use `:inputs_wrapper=>:bs3` when registering the configuration.

I probably won't be able to do a full review on this until about a week from now.

**Comment by kematzy** [comment]
Hi Jeremy,

Apologies for prolonged silence on this, as I have been away on 3-weeks
holiday in UK/Sweden.

I have pushed some work-in-progress stuff and would really appreciate some
feedback and some help over some stumbling blocks.

I'm sorry to admit that my Ruby fu is very far from yours and I have
therefore got stuck on some of your logic / code flow.

I understand that you are busy too, so I will patiently await any reply.

Kind regards,

Kematzy


-- 
Kind regards,

Mats Persson
E: kematzy@gmail.com
T: +60 12 - 2345 194

**Comment by jeremyevans** [comment]
I'll try out your fork soon and see if I can get more of the tests passing.  Thanks for your continued work on this.

**Comment by jeremyevans** [comment]
Looks like a lot of the issues stem from trying to introspect the `tag` argument in the transformers instead of the `input` argument.  Here's a patch that should get it passing with only 14 skipped specs: http://pastie.org/10367008

If you have more questions about a specific area of the code, please let me know.

**Comment by kematzy** [comment]
Hi Jeremy,

A very big Thank You! for your review and patch.

I have incorporated the patch and fixed most issues, but have a few ones
left which appears to be beyond my understanding / skills :-(

I have skipped those tests and added comments to highlight the issues.

My sincere apologies for the messy code in the serializer.rb file, I'm sure
it can be cleaned up drastically, but I was trying to hunt down the
@f.tag(:textarea) issue without much success.

Thanks for your time and attention.




-- 
Kind regards,

Mats Persson
E: kematzy@gmail.com
T: +60 12 - 2345 194

**Comment by jeremyevans** [comment]
Looks like things were pretty much already ready.  Form#tag is a low-level interface, it's not going to use any of the transformers, so I deleted one spec that expected that, and just removed the skips from the other specs since they looked good.  There was one ruby 1.8 compatibility issue which I fixed as well, as well as reverting an unnecessary change to spec/forme_spec.rb (which was probably added to work around buggy editor syntax highlighting).  Here's a link to the patch: http://pastie.org/pastes/10368391/text

Assuming that you are happy with things and don't want more changes, you should delete the commented out debugging code, and squash the bs3-support into a single commit, then submit a pull request for that commit.  If you don't feel comfortable doing the squashing, just submit the pull request after deleting the debugging code and I'll squash it when I merge.

Thanks for your continued help on this!

**Comment by badosu** [comment]
Hi @kematzy how is the status of your BS3 branch?

If you have simple setup instructions I may use it and even collaborate on eventual holes.

**Comment by kematzy** [comment]
Hi Amadeus,

I was trying to update the demo-site code with some BS3 examples when I
discovered some Sequel plugin & BS3 output issues that I'm working through.

I'm hoping to push my work soon and following that, I hope Jeremy will have
time to review, approve, squash git commits and finally include BS3 in the
official gem.

As for simple setup instructions:

-- to help out improving the code, just wait for my push and then pull repo
and run "bundle install" and go at it.

-- to use Forme with BS3 support in your app:  again wait for my push and
then vendor (include in app) the code and just add ":config=>:bs3" to your
form tag or new Forme::Form instances. See spec's for guidance. Please note
that the BS3 code may not be production ready!


Collaboration is most welcome and the main holes are related to handling:

 --  BS3 skipped tests and TODO's outlined in the pushed code.
 --  BS3 input groups (.input-group, etc)
 --  BS3 form alternatives (inline & horizontal, etc)
 --  BS3 validations (.has-success, .has-warning, etc)
 --  BS3 feedback icons (.has-feedback, etc)

Regards,

Kematzy



-- 
Kind regards,

Mats Persson
E: kematzy@gmail.com
T: +60 12 - 2345 194

**Comment by kematzy** [comment]
Hi Jeremy,

I think I have done as much as I can do at this stage and would really appreciate your input.

There are 13 skipped (failing) tests that I just don't know how handle the required code changes for, so I'm hoping you are able to overcome those issues.

As highlighted above there are still some outstanding BS3 support items, but around 80%+ is there.

I have also taken a stab at updating / improving the demo-site with the new BS3 support. Hopefully I have not changed things too much.

I'm not very sure on how to squash all the changes into a single commit, so I leave that for you.

Please feel free to amend as required before possibly merging my work into HEAD.

Thanks a lot for your time, attention and assistance throughout this. Very much appreciated.

**Comment by jeremyevans** [comment]
OK.  I'll take another look at your branch, but it might be a few days before I have time.  Thanks again for your help.

**Comment by jeremyevans** [comment]
I got the first skipped spec working, here's the diff:  http://pastie.org/pastes/10382656/text 

This diff changes the output for a lot of the specs in forme_bs3_spec, and I'm not sure if those changes are desired or not, so you'll have to check them.  If the changes are correct, you'll need to fix those specs.  If the changes are not correct, please let me know why and we can discuss how to fix it.

Please see if you can fix the rest of the specs with this code change.  If you still have issues you can't fix, please let me know here.  Please don't add skipped specs with commented out code, since it was difficult for me to figure out what you actually wanted to pass.  Just leave the specs as failing and don't have commented code in the specs so I know what to fix.  If forme is failing to infer the options as you think it should, it may be good to follow the example I used in the fixed spec, where first you test with the explicit options, and next you test without the explicit options.  That way it is obvious to me that the problem is the inability of forme to infer the correct options.

**Comment by badosu** [comment]
@kematzy I tried to use the :bs3 configuraton but I received a weird error.

**Comment by kematzy** [comment]
Hi Jeremy, 

Thanks for your patch and assistance, which corrected most of the issues.

There are now **only three (3) failing tests remaining** to resolve, all of them relating to the `:checkboxset` / `:radioset` functionality when an error is present.

A single `:checkbox` / `:radio` or a `:checkboxset` / `:radioset` with an error should be wrapped in a 
`<div class="has-error">...</div>` tag. (see examples below)

Expected output format should be as follows:

A single input with error: 

    <div class="has-error">
      
      <div class="checkbox/radio">
        <label>
          <input ...>
        </label>
      </div>
      
      <span class="help-block with-errors">Message...</span>
    </div>


A set of checkbox / radio inputs

    <div class="has-error">
      
      <div class="checkbox/radio">
        <label>
          <input ...>
        </label>
      </div>
      
      <div class="checkbox/radio">
        <label>
          <input ...>
        </label>
      </div>
      
      <span class="help-block with-errors">Message...</span>
    </div>

I have tried many hacks to try to resolve this issue, but without success.

The problem seems to be in the `formatter.rb` file **around lines 250**, where the error and error message is tagged onto the last input (checkbox/radio).

I hope I have improved my notes on failing specs so that it's clear what should be output and what the current error is. All BS3 tests should now have the correct expected output format.

You can search for `# TODO: INCORRECT OUTPUT: Missing 'has-error' on :wrapper tag` to identify the failing specs. 

If this remaining issue can be resolved then all tests will pass and I will consider most of my work done.

Again, I thank you for your time and attention.

**Comment by kematzy** [comment]
@badosu Could you please share some more information about your "weird error"??

For me the BS3 support is working generally well and without any major issues. I'm sure there's still some issues that I might not have identified that will be cropping up.

**Comment by jeremyevans** [comment]
@kematzy Thanks, I'll try to do another review of these changes within a few days, and see if I can get the remaining tests passing.

**Comment by badosu** [comment]
@kematzy Using `:bs3` config gave me `nil can't be a Serializer` issue.

I am using my own configuration now, this is making me learn how to configure `Forme` properly, as soon as I can I'll compare it to your work and see if I can help with anything,

**Comment by jeremyevans** [comment]
Here's a diff that fixes the 3 remaining failures: http://pastie.org/pastes/10396100/text

However, because it requires adding a div wrapper by default, it currently breaks 33 existing bs3 specs.  I'm not sure if you want the wrapping div only to be added by default for inputs with errors (I doubt it), but my guess is the specs need to be updated.  If you really just want to add a wrapping div when the radioset/checkboxset has an error and not when it doesn't have an error, that's possible, you should just add a new wrapper transformer that only wraps with div in case of an error.

Hopefully with the patch, you can get the rest of the specs passing to your satisfaction.  If you run into more issues, please let me know.

**Comment by kematzy** [comment]
I have updated the code again and all tests are passing and the output is expected and correct in my mind.

The error output for `:checkboxset` / `:radioset` is not clearly defined in Bootstrap, which is why I kind of created my own interpretation.

To assist you, and others, to visualise the differences I have created a full BS3 HTML reference document (see <tt>spec/bs3_reference.html</tt>) which clearly displays the differences and how error handling works for each form element.

If there are objections to my error handling output for `:checboxset` and `:radioset` then I'm OK with it being changed.

I have created a new [branch](https://github.com/kematzy/forme/tree/add-bs3-support) with less and cleaner commits for you to merge.

Hopefully this is OK and can be merged in quickly.

Thanks a lot for your assistance and patience.
